bam_consensus.c

```diff
@@ -0,0 +1,1702 @@
+
+        // Correction for mapping quality.  Maybe speed up via lookups?
+        // Cannot nullify mapping quality completely.  Lots of (true)
+        // SNPs means low mapping quality.  (Ideally need to know
+        // hamming distance to next best location.)
+
+        if (flags & CONS_MQUAL) {
+            int mqual = b->core.qual;
+            if (opts->nm_adjust) {
+                mqual /= (nm_local(p, b, pos)+1);
+                mqual *= 1 + 2*(0.5-(td>30?30:td)/60.0); // depth fudge
+            }
+
+            // higher => call more; +FP, -FN
+            // lower  => call less; -FP, +FN
+            mqual *= opts->scale_mqual;
+
+            // Drop these?  They don't seem to ever help.
+            if (mqual < opts->low_mqual)
+                mqual = opts->low_mqual;
+            if (mqual > opts->high_mqual)
+                mqual = opts->high_mqual;
+
+            double _p = 1-q2p[qual];
+            double _m = mqual_pow[mqual];
+            qual = ph_log(1-(_m * _p + (1 - _m)/4)); // CURRENT
+            //qual = ph_log(1-_p*_m); // testing
+            //qual *= 6/sqrt(td);
+        }
+
+        /* Quality 0 should never be permitted as it breaks the maths */
+        if (qual < 1)
+            qual = 1;
+
+        __ = p__[qual];       // neither match
+        MM = pMM[qual] - __;  // both match
+        _M = p_M[qual] - __;  // one allele only (half match)
+
+        if (flags & CONS_DISCREP) {
+            qe = q2p[qual];
+            sumsC[base] += 1 - qe;
+        }
+
+        counts[base]++;
+
+        switch (base) {
+        case 0: // A
+            S[0] += MM;
+            S[1] += _M;
+            S[2] += _M;
+            S[3] += _M;
+            S[4] += _M;
+            break;
+
+        case 1: // C
+            S[1] += _M;
+            S[5] += MM;
+            S[6] += _M;
+            S[7] += _M;
+            S[8] += _M;
+            break;
+
+        case 2: // G
+            S[ 2] += _M;
+            S[ 6] += _M;
+            S[ 9] += MM;
+            S[10] += _M;
+            S[11] += _M;
+            break;
+
+        case 3: // T
+            S[ 3] += _M;
+            S[ 7] += _M;
+            S[10] += _M;
+            S[12] += MM;
+            S[13] += _M;
+
+            break;
+
+        case 4: // *
+            S[ 4] += _M;
+            S[ 8] += _M;
+            S[11] += _M;
+            S[13] += _M;
+            S[14] += MM;
+            break;
+
+        case 5: /* N => equal weight to all A,C,G,T but not a pad */
+            S[ 0] += MM;
+            S[ 1] += MM;
+            S[ 2] += MM;
+            S[ 3] += MM;
+            S[ 4] += _M;
+
+            S[ 5] += MM;
+            S[ 6] += MM;
+            S[ 7] += MM;
+            S[ 8] += _M;
+
+            S[ 9] += MM;
+            S[10] += MM;
+            S[11] += _M;
+
+            S[12] += MM;
+            S[13] += _M;
+            break;
+        }
+
+        depth++;
+
+        if (p->eof && p->cd)
+            free(p->cd);
```

Comment:
It might be a good idea to set this to NULL after free.

Reply:
It's harmless, as it's initialised to zero when they're reused (https://github.com/jkbonfield/samtools/blob/consensus_part1b/consensus_pileup.c#L494).  However it's not really detrimental so can add it.

Arguably a better interface would be to not free it and cache the length it was allocate to, so it doesn't need reallocating, but that's quite a bit more to do and it would need a tidy up function at the end too.